src/EventSubscriber/SearchNoHitEventSubscriber.php

```diff
@@ -57,17 +64,101 @@ public static function getSubscribedEvents(): array
+                $keyedNoHits[$cacheKey] = $noHit;
+            }
+
+            $nonCommittedCacheItems = $this->getNonCachedNoHits($keyedNoHits);
+            $this->sendSearchNoHitEvents($nonCommittedCacheItems);
+        }
+    }
+
+    /**
+     * Send search no hit events.
+     *
+     * @param array $nonCommittedCacheItems
+     *   Array of cache items
+     */
+    private function sendSearchNoHitEvents(array $nonCommittedCacheItems): void
+    {
+        foreach ($nonCommittedCacheItems as $cacheItem) {
+            /** @var NoHitItem $noHitItem */
+            $noHitItem = $cacheItem->get();
+            $message = new ProcessMessage();
+            $message->setIdentifierType($noHitItem->getIsType())
+                ->setIdentifier($noHitItem->getIsIdentifier());
+
+            $this->noHitsCache->saveDeferred($cacheItem);
+
+            $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        }
 
-                $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        $this->noHitsCache->commit();
+    }
+
+    /**
+     * Get cache items for the identifiers not present in the cache.
+     *
+     * @param array $keyedNoHits
+     *   Array og cacheKey => NoHitItem pairs
+     *
+     * @return array
+     *   Array of cache items not yet committed to cache
+     */
+    private function getNonCachedNoHits(array $keyedNoHits): array
+    {
+        $nonCommittedCacheItems = [];
+        try {
+            $cacheKeys = array_keys($keyedNoHits);
+            $cacheItems = $this->noHitsCache->getItems($cacheKeys);
+            foreach ($cacheItems as $cacheItem) {
+                if (!$cacheItem->isHit()) {
+                    /** @var NoHitItem $noHitItem */
+                    $cacheKey = $cacheItem->getKey();
+                    $noHitItem = $keyedNoHits[$cacheKey];
+                    $cacheItem->set($noHitItem);
+                    $nonCommittedCacheItems[] = $cacheItem;
+                }
             }
+        } catch (InvalidArgumentException $e) {
+            // @TODO Logging?
+        }
+
+        return $nonCommittedCacheItems;
+    }
+
+    /**
+     * Get an array of valid cache keys for the identifiers.
+     *
+     * Keys should only contain letters (A-Z, a-z), numbers (0-9) and the _ and . symbols.
+     *
+     * @see https://www.php-fig.org/psr/psr-6/
+     * @see https://symfony.com/doc/current/components/cache/cache_items.html#cache-item-keys-and-values
+     *
+     * @param string $type
+     *   The identifier type
+     * @param string $identifier
+     *   The identifier
+     *
+     * @return string
+     *   The cache key
+     */
+    private function getValidCacheKey(string $type, string $identifier): string
+    {
+        if (IdentifierType::PID === $type) {
+            $identifier = str_replace(':', '_', $identifier);
+            $identifier = str_replace('-', '_', $identifier);
         }
+
+        return $key = $type.'.'.$identifier;
```

Comment:
Why `$key =` here in the return statement?

Reply:
Copy/Paste error